### PR TITLE
Add support for gecko-only array values; use for stylo

### PR DESCRIPTION
**Pull request by Manishearth**
Doesn't yet work, because I can't grow nsTArrays from Rust. If anyone knows how to add bindings for that, let me know!

Thoughts on the design so far? Once this PR lands, all of the array-accepting background- properties can just use gecko_autoarray_longhand and some iterators in the geckolib implementation without changing much other code to work with arrays.

cc @emilio @bholley

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11851)

### CODE REVIEW

File: ports/geckolib/properties.mako.rs [hunk 1]
```diff
@@ -853,76 +853,84 @@ fn static_assert() {
...
-                    };
-
-                    // TODO: figure out what gecko does in the `corner` case.
-                    if let AngleOrCorner::Angle(angle) = gradient.angle_or_corner {
-                        unsafe {
-                            (*gecko_gradient).mAngle.set(angle);
+        self.gecko.mImage.mImageCount = cmp::max(self.gecko.mImage.mLayers.len() as u32,
+                                                 self.gecko.mImage.mImageCount);
+
+        // TODO: pre-grow the nsTArray to the right capacity
```
Comment (Manishearth) [inline ports/geckolib/properties.mako.rs]: This is where I need the bindings

File: ports/geckolib/properties.mako.rs [hunk 2]
```diff
@@ -853,76 +853,84 @@ fn static_assert() {
...
+        // otherwise the below code won't work
+        for (image, geckoimage) in images.0.into_iter().zip(self.gecko.mImage.mLayers.iter_mut()) {
```
Comment (Manishearth) [inline ports/geckolib/properties.mako.rs]: Github doesn't show this diff well, but the below code is simply indented one level with a change to the `Gecko_SetGradientImageValue` line

File: components/style/properties/helpers.mako.rs [hunk 3]
```diff
@@ -33,6 +33,81 @@
     </%call>
 </%def>
 
+<%doc>
+    To be used in cases where we have a grammar like
+    "<thing> [ , <thing> ]*", but only support a single value
+    in servo
+</%doc>
+<%def name="gecko_autoarray_longhand(name, **kwargs)">
```
Comment (Manishearth) [inline components/style/properties/helpers.mako.rs]: once servo starts supporting `<thing> [ , <thing> ]*` properties we can refactor this to work in general and have a thin conditional-compile wrapper around it.

**Comment by bholley** [comment]
Nice! In terms of reallocating, I think we should just do an FFI call, otherwise there's lots of stuff we need to worry about.

Didn't look too closely here, @emilio should probably review since he did the initial array stuff. But it's great to see a clean and easy abstraction on this stuff! :-)

**Comment by emilio** [comment]
Yeah, reallocating is basically why I didn't made a more vec-like interface for nsTArray.

I can't probably take a deeper look today (I'm packing things), but yeah, I'd go with the FFI call to reallocate. 

Probably we could share allocators (I already had a patch for it, though I need to double-check it does the correct thing evrrywhere), but for now I think we should try to avoid duplicating all the logic there.

If a template function with a pointer to the nsTArray doesn't work (I don't think it'd work, since the functions won't ever be instantiated), probably passing the size of the element would be sufficient, though it'd require to do a bit more math.

Probably receiving a size and a pointer to nsTArray<uint8_t>, and reallocate to n * size elements could work.

I can prioritize the allocator patch though, if you'd prefer to try to write that logic in rust.

**Comment by emilio** [comment]
nsTArray<uint8_t>, I meant

**Comment by emilio** [comment]
Ugh, github mobile strips my less than signs, I meant an nsTArray of uint8_t

File: components/style/properties/helpers.mako.rs [hunk 4]
```diff
@@ -33,6 +33,81 @@
...
+            }
+            pub mod computed_value {
+                use super::single_value;
+                #[derive(Debug, Clone, PartialEq, HeapSizeOf)]
+                pub struct T(pub Vec<single_value::computed_value::T>);
+            }
+
+            impl ToCss for computed_value::T {
+                fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+                    if self.0.len() >= 1 {
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: This is probably a bit more tricky. In some props we might need to write `none` in the empty case, eg, animation-name.

File: components/style/properties/helpers.mako.rs [hunk 5]
```diff
@@ -33,6 +33,81 @@
...
+            }
+            pub fn get_initial_value() -> computed_value::T {
+                computed_value::T(vec![single_value::get_initial_value()])
+            }
+            pub fn parse(context: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
+                input.parse_comma_separated(|parser| {
+                    single_value::parse(context, parser)
+                }).map(|ok| SpecifiedValue(ok))
+            }
+            impl ToComputedValue for SpecifiedValue {
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: Some (most?) of these properties are computed-value as specified, which would just involve a generic clone which would be cheaper.

I'm not opposed to land this as is, but I'd want a note about that here.

**Comment by bholley** [comment]
I think you can force-instantiate a generic function in C++. Something like http://stackoverflow.com/a/2155790/3670407 in the cpp file.

**Comment by bholley** [comment]
Oh I guess the issue here is that we don't necessarily know the type we want it for in C++? I guess in that case just passing a size might be better.

**Comment by Manishearth** [comment]
Is the allocator patch necessary? I was hoping we could call whatever grow function gecko uses through FFI; I just don't yet know which function that is (havent really looked), and what button to press to get bindings for it.

That way we can use whatever reallcoation logic that gecko uses. And we are not dependent on shared allocators.

**Comment by Manishearth** [comment]
Oh, monomorphization. Ugh.

**Comment by emilio** [comment]
Oh, sure you can force-instantiate it, and we can probably know most if not all of the types we'd need. I don't want to know what MSVC would think about that though :P

So yea, I'd go for an ffi fn like:

```
void Gecko_ResizeArray(nsTArray<uint8_t>* arr, uint32_t n, uint32_t element_size);
```

**Comment by bholley** [comment]
If there are only a couple of types, it might be nicer to just have explicit separate FFI calls for each T to avoid too much scary munging. Is it more like 3 or more like 15?

**Comment by emilio** [comment]
I think there were about 6 different types in the layers structure, and a few more at least.

**Comment by Manishearth** [comment]
So, https://dxr.mozilla.org/mozilla-central/source/xpcom/glue/nsTArray.h#380 exists and is really all we need here. It is not parametrized on the type, just on the ActualAlloc (not sure what actualalloc we use, or what exactly that is). So it will always be monomorphized.

How do I get bindings to this?

**Comment by emilio** [comment]
@Manishearth So our `nsTArray` bindings are bindings to an actual simpler *replacement* of `nsTArray` (search for `nsTArray_Simple`.

That doesn't change the way we can generate bindings to it though, and it'd be creating a gecko function with the signature as above, and just calling that other function from gecko instead of doing it manually.

**Comment by Manishearth** [comment]
I'm having tons of trouble building bindgen.

First, I tried using llvm38, getting the following error:

```rust
"_clang_Cursor_isFunctionInlined", referenced from:
bindgen::clang::Cursor::is_inlined_function::hbf3ecb6d1f288ea8 in libbindgen.rlib(bindgen.0.o)
```

I have a local clone of LLVM that is close to trunk (cloned a month ago), which I installed. This made the bindgen build work.

Then when running bindgen I get `obj-x86_64-apple-darwin15.3.0//dist/include/mozilla/mozalloc.h:16:12: fatal error: 'new' file not found`. If I include `m-c/build/stlport/stlport/`, that goes away and I get a  similar error for `cstddef`. Fixing that by explicitly including `/usr/include/c++/4.2.1/` seems to get rid of all the include errors, but  now I have errors like:

```
unsupported type `?` (obj-x86_64-apple-darwin15.3.0//dist/include/mozilla/ServoBindings.h:37:31)
unsupported type `?` (obj-x86_64-apple-darwin15.3.0//dist/include/mozilla/ServoBindings.h:174:36)
```

I'm pretty sure I've done something wrong at this point 😄 

Could someone with a working bindgen setup explain how theirs was set up? Preferably on a mac.

**Comment by bholley** [comment]
Yeah, you need to brew install llvm38. We don't actually need the tip stuff for stylo, so @emilio is going to make that dep optional.

For the headers, you need to make sure your standard library is installed to /usr/local. On mac, I think xcode-select --install should do it.

**Comment by Manishearth** [comment]
I did run xcode-select --install, it seemed to already exist. Mine are in /usr/include though. They work when used on an external C++ file; just seems like bindgen can't find them.

So currently we need to use llvm tip to get things working, right? llvm38 is just a future goal?

**Comment by bholley** [comment]
Sorry, I meant /usr/include. Should work if that's the case. Are you running the scripts in gecko_bindings/tools?

Currently llvm38 is what you need, llvm-stable is a future goal.

**Comment by Manishearth** [comment]
Okay, so llvm-38 doesn't work for me (it gives that link error), and the usr/include stuff doesn't help either. I am running the scripts from the tools folder.

**Comment by Manishearth** [comment]
So I got rid of the make install'd llvm by make installing _again_ and deleting everything it installed. I then reinstalled llvm38. Things work now, so this was all my fault 😄

**Comment by Manishearth** [comment]
Not done yet, but the bindings are ready

**Comment by Manishearth** [comment]
Still doesn't run, but it wasn't running before I made this PR either. Not sure if the crash is the same:

```
#01: mozilla::ServoStyleSet::ResolveAnonymousBoxStyle(nsIAtom*, nsStyleContext*, unsigned int)[/Users/manishearth/mozilla/quantum-central/obj-x86_64-apple-darwin15.3.0/dist/NightlyDebug.app/Contents/MacOS/XUL +0x4bc9a63]
#02: mozilla::StyleSetHandle::Ptr::ResolveAnonymousBoxStyle(nsIAtom*, nsStyleContext*, unsigned int)[/Users/manishearth/mozilla/quantum-central/obj-x86_64-apple-darwin15.3.0/dist/NightlyDebug.app/Contents/MacOS/XUL +0x4e1d447]
#03: nsCSSFrameConstructor::ConstructTable(nsFrameConstructorState&, nsCSSFrameConstructor::FrameConstructionItem&, nsContainerFrame*, nsStyleDisplay const*, nsFrameItems&)[/Users/manishearth/mozilla/quantum-central/obj-x86_64-apple-darwin15.3.0/dist/NightlyDebug.app/Contents/MacOS/XUL +0x4dd1871]
#04: nsCSSFrameConstructor::ConstructFrameFromItemInternal(nsCSSFrameConstructor::FrameConstructionItem&, nsFrameConstructorState&, nsContainerFrame*, nsFrameItems&)[/Users/manishearth/mozilla/quantum-central/obj-x86_64-apple-darwin15.3.0/dist/NightlyDebug.app/Contents/MacOS/XUL +0x4dda577]
#05: nsCSSFrameConstructor::ConstructFramesFromItem(nsFrameConstructorState&, nsCSSFrameConstructor::FrameConstructionItemList::Iterator&, nsContainerFrame*, nsFrameItems&)[/Users/manishearth/mozilla/quantum-central/obj-x86_64-apple-darwin15.3.0/dist/NightlyDebug.app/Contents/MacOS/XUL +0x4ddeeea]
```

File: components/style/properties/helpers.mako.rs (hunk 4 above)
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: Oh, also: !self.0.is_empty()

**Comment by Manishearth** [comment]
I needed to call Initialize(). Obviously ;)

**Comment by Manishearth** [comment]
Looks ugly, but it's correct! :

screen shot 2016-07-06 at 6 18 10 pm

**Comment by emilio** [comment]
I'm reviewing one revision at a time, so please ignore any comment about code that is gone/fixed in later revisions.

---

The approach here looks good, though I left some nits and questions through the code :)

BTW, I'm not sure if I've thanked you for doing this, but doesn't hurt doing it again :)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, 2 of 2 files at r6, 3 of 3 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, 8 of 8 files at r10, 2 of 2 files at r11, 2 of 2 files at r12, 3 of 3 files at r13, 3 of 3 files at r14.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[components/style/properties/helpers.mako.rs, line 79 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11851#-KM1UzBNpP3i6CT5ozYK:-KM1UzBNpP3i6CT5ozYL:255360083) ([raw file](https://github.com/servo/servo/blob/12c50dc4b6a4d1866a3b1565c6efb5c10573f73a/components/style/properties/helpers.mako.rs#L79)):*
> ```Rust
>             impl ToCss for SpecifiedValue {
>                 fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
>                     if self.0.len() >= 1 {
> ```

Ditto here re `is_empty`, `none` and `ComputedValueAsSpecified`, we probably want to have some kind of argument in the mako template for that.

---

*[ports/geckolib/properties.mako.rs, line 958 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11851#-KM1WjGjMiuWsun4O-KA:-KM1WjGjMiuWsun4O-KB:608351262) ([raw file](https://github.com/servo/servo/blob/f2ce4a5883a076d9fec722a7eeaf9331631642ab/ports/geckolib/properties.mako.rs#L958)):*
> ```Rust
>         unsafe {
>             // Prevent leaking of the last element we did set
>             for image in &mut self.gecko.mImage.mLayers {
> ```

nit: put the unsafe block before the loop?

---

*[ports/geckolib/gecko_bindings/bindings.rs, line 229 \[r7\]](https://reviewable.io:443/reviews/servo/servo/11851#-KM1WEFmdXuGUs06tFke:-KM1WEFmdXuGUs06tFkf:1245752056) ([raw file](https://github.com/servo/servo/blob/2f5120e1e79fadc699e8de3e95b6f6b510ae8400/ports/geckolib/gecko_bindings/bindings.rs#L229)):*
> ```Rust
>                                     src: *const nsStyleDisplay);
>     pub fn Gecko_ArrayEnsureCapacity(array: *mut ::std::os::raw::c_void,
>                                      capacity: usize, size: usize);
> ```

Is there already a relevant gecko bug for this?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_auto_array.rs, line 1 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11851#-KM1VZH13yC520VP5FTn:-KM1VZH13yC520VP5FTo:81624155) ([raw file](https://github.com/servo/servo/blob/a070901f96661f7760bf4ea9f5d2aeb95fb987bd/ports/geckolib/gecko_bindings/sugar/ns_style_auto_array.rs#L1)):*
> ```Rust
> use std::iter::{once, Chain, Once, IntoIterator};
> ```

Tidy is going to complain about the lack of license here :)

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 38 \[r7\]](https://reviewable.io:443/reviews/servo/servo/11851#-KM1WJQPFqplkUGzMs7q:-KM1WJQQpS1y512-qMfm:218462260) ([raw file](https://github.com/servo/servo/blob/2f5120e1e79fadc699e8de3e95b6f6b510ae8400/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L38)):*
> ```Rust
>         unsafe { mem::transmute(self.mBuffer) }
>     }
>     fn header_mut <'a>(&'a mut self) -> &'a mut nsTArrayHeader {
> ```

nit: no space before `<'a>`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11851#-:-KM1VJvFEhHgQ-FspCGJ:838553802)*

**Comment by Manishearth** [comment]
Done. Bug filed at https://bugzilla.mozilla.org/show_bug.cgi?id=1285235

**Comment by Manishearth** [comment]
Updated with changes from bugzilla review

**Comment by emilio** [comment]
Apart from these comments, everything else looks good to me. Let me know if there's anything about any of those comments that might be unclear.

Probably @SimonSapin knows a lot already about how different list values might be handled.

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 9 files at r15.
Review status: 8 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/style/properties/helpers.mako.rs, line 41 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRjnlJsEB5MZtYXqwt:-KMRjnlJsEB5MZtYXqwu:-1227292989) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/components/style/properties/helpers.mako.rs#L41)):*
> ```Rust
>     in servo
> </%doc>
> <%def name="gecko_autoarray_longhand(name, **kwargs)">
> ```

I think we still want a few more arguments here.

First, we probably want a fast path for the "computed value as specified case", so an arg here like `computed_value_as_specified=False`, and then skipping all the re-declaration of the computed values and just do `impl ComputedValueAsSpecified for computed_value::T {}`.

Other thing we need to look at is the handling of the empty list in the `ToCss` trait. I think the most common case is "if there's no specified value, just spit `none`", but we'll need to see how many different handling is here, and allowing to switch depending on the property.

It's fine for me if you want to leave it as a followup (though I'd rather not), but that's something eventually we'll need to take care about, and at least a `FIXME` comment so we don't forget about it would be needed.

---

*[ports/geckolib/properties.mako.rs, line 964 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRlcvDFB53Ezu_0-wy:-KMRlcvDFB53Ezu_0-wz:-209834743) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/properties.mako.rs#L964)):*
> ```Rust
>             Gecko_EnsureImageLayersLength(&mut self.gecko.mImage, images.0.len());
>             for image in &mut self.gecko.mImage.mLayers {
>                 Gecko_InitializeImageLayer(image, Background as u8);
> ```

I'd kind of prefer something like:

```rust
use gecko_bindings::structs::nsStyleImageLayers_LayerType as LayerType;
...
Gecko_InitializeImageLayer(image, LayerType::Background as u8);
```

But it's fine too if you prefer to keep as-is.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_auto_array.rs, line 14 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRntHrm4v7i76oMUH1:-KMRntHrm4v7i76oMUH2:-1842979210) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/gecko_bindings/sugar/ns_style_auto_array.rs#L14)):*
> ```Rust
>     }
> 
>     pub fn len(&self) -> usize {
> ```

Also, could you add a warning here saying that normally there's another field in the structure indicating the actual length (that might be zero)?

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 38 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRm_iviiqyv0iUTUbk:-KMRm_iviiqyv0iUTUbl:-1055157307) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L38)):*
> ```Rust
>         unsafe { mem::transmute(self.mBuffer) }
>     }
>     fn header_mut<'a>(&'a mut self) -> &'a mut nsTArrayHeader {
> ```

Hmm... I think this should be unsafe, since the `nsTArrayHeader` in the empty case might be a pointer to a C++ static variable, so modifying that could make nasty stuff happen.

This shouldn't be a problem as long as we use it the way we use it, but can you leave a comment here about this?

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 49 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRnYBicSYHY-UkAJG7:-KMRnYBicSYHY-UkAJG8:1211766557) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L49)):*
> ```Rust
>     }
> 
>     fn ensure_capacity(&mut self, cap: usize) {
> ```

I'm unsure why the whole `u32 -> usize -> u32` roundtrip. If gecko just uses `u32` we can probably stick with that here too (if the relevant gecko patch hasn't landed yet of course).

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 56 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRn4PJLJyqWbB_GaeW:-KMRn4PJLJyqWbB_GaeX:-31501653) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L56)):*
> ```Rust
> 
>     // unsafe because the array may contain uninits
>     pub unsafe fn set_len(&mut self, len: u32) {
> ```

I know for now we don't deal with this kind of stuff, but could you `debug_assert!` that `len >= self.len()`. I don't want to accidentally leak values in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11851#-:-KMRmCjvKCCAfY85_ujS:81686603)*

**Comment by Manishearth** [comment]
udpated

**Comment by emilio** [comment]
LGTM with these nits and the gecko-side changes :)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 9 files at r15, 5 of 6 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/style/properties/helpers.mako.rs, line 36 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMVgaNl6PTPMfpiVOWe:-KMVgaNl6PTPMfpiVOWf:-438009739) ([raw file](https://github.com/servo/servo/blob/aa1526d986347aa23eab3053e6d9c85743b1e6dc/components/style/properties/helpers.mako.rs#L36)):*
> ```Rust
> </%def>
> 
> // FIXME (Manishearth): Add computed_value_as_specified argument and handle null case
> ```

nit: "Handle correctly the empty case" may be more understandable?

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 49 \[r15\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMRnYBicSYHY-UkAJG7:-KMVhVxEfDWXYPHOeQkP:-1525609190) ([raw file](https://github.com/servo/servo/blob/8f7dcff1ff9f5bbfececcd8022f5e2683cf93fbe/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L49)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> I'm unsure why the whole `u32 -> usize -> u32` roundtrip. If gecko just uses `u32` we can probably stick with that here too (if the relevant gecko patch hasn't landed yet of course).
</details>
FTR, gecko uses `size_t` in its public API, yet they store the length as a `uint32_t`, so there isn't much we can do about that for now, and that should probably belong to another patch.

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 38 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMVhhYOXzbI4Hq4qfNr:-KMVhhYPV2fYT6sj50_C:-1540430903) ([raw file](https://github.com/servo/servo/blob/aa1526d986347aa23eab3053e6d9c85743b1e6dc/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L38)):*
> ```Rust
>         unsafe { mem::transmute(self.mBuffer) }
>     }
>     // unsafe, since header may be in shared static or something
> ```

"may be a pointer to a shared static"?

---

*[ports/geckolib/gecko_bindings/tools/regen.py, line 126 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMVhG9TmSTRU7FFSMkb:-KMVhG9TmSTRU7FFSMkc:1290844304) ([raw file](https://github.com/servo/servo/blob/aa1526d986347aa23eab3053e6d9c85743b1e6dc/ports/geckolib/gecko_bindings/tools/regen.py#L126)):*
> ```Python
>             "FontFamilyType", "nsIAtom",
>         ],
>         "void_types": [
> ```

Hmm... I'd expect these to be generated by default. I can certainly stand with it as a hack for now, but I'll need to look into it after this lands.

---

*[ports/geckolib/gecko_bindings/tools/regen.py, line 249 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11851#-KMVh-OJYPrccX-SszuZ:-KMVh-OJYPrccX-Sszu_:-1025933169) ([raw file](https://github.com/servo/servo/blob/aa1526d986347aa23eab3053e6d9c85743b1e6dc/ports/geckolib/gecko_bindings/tools/regen.py#L249)):*
> ```Python
>             # TODO: this is hacky, figure out a better way to do it without
>             # hardcoding everything...
>             if ty.startswith("nsStyle"):
> ```

Can you change this for `if ty_fragments[-1].startswith("nsStyle")`? We certainly do not need to implement Send and Sync in nested classes by default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11851#-:-KMVi2YzzBrJyXO57zsQ:241200947)*

**Comment by emilio** [comment]
@bors-servo: r+

Awesome! :)

**Comment by emilio** [comment]
@bors-servo: p=1

 * It makes the repo in sync for stylo development.

**Comment by Manishearth** [comment]
> First, we probably want a fast path for the "computed value as specified case", so an arg here like computed_value_as_specified=False, and then skipping all the re-declaration of the computed values and just do impl ComputedValueAsSpecified for computed_value::T {}.

This doesn't actually work in this case, because the values aren't the same. One is `computed::Image`, the other is `specified::Image` (they differ because the linear gradients contain lengths which are represented differently once computed)

**Pull request merged**